src/mame/sega/sega_beena.cpp

```diff
@@ -71,131 +102,2074 @@
+		uint16_t tile_transform = (tile & 0xf000) >> 12;
+		uint16_t flip_x = (tile_transform & 1) != 0;
+		uint16_t flip_y = (tile_transform & 2) != 0;
+		bool is_overlay = (tile_transform & 4) != 0;
+		if (is_overlay != is_overlay_rendered) {
+			continue;
+		}
+
+		uint16_t tile_idx = tile & 0x7ff;
+		bool is_sprite_tile_data_used = tile & 0x800;
+		gfx_element *gfx = m_gfxdecode->gfx(is_sprite_tile_data_used ? 1 : 0);
+		const pen_t *paldata = m_palette->pens() + gfx->colorbase() + gfx->granularity() * (7 % gfx->colors());
+		const uint8_t scale = m_scale;
+		drawgfxzoom_with_pixel_op(gfx, bitmap, cliprect,
+			tile_idx,
+			flip_x,
+			flip_y,
+			x * m_scale,
+			y * m_scale,
+			0x10000 * m_scale,
+			0x10000 * m_scale,
+			[scale, paldata](int32_t y, const std::unique_ptr<rgb_t[]> &palcache, uint32_t &destp, const uint8_t &srcp) {
+				if (srcp != 0) {
+					rgb_t color = palcache[y / scale * 0x100 + srcp];
+					if (color == 0) {
+						color = paldata[srcp];
+					}
+					destp = color;
+				}
+			});
+	}
+}
+
+/**
+ * Similar to gfx_element::drawgfxzoom_core(), but uses a pixel_op that depends on the current y-position to render,
+ * in order to apply dedicated palettes for each scanline. These palettes are cached and passed directly
+ * as a parameter instead of being captured in the pixel_op lambda, to avoid degraded performance from large copies.
+ */
+template <typename BitmapType, typename FunctionClass>
+inline void sega_9h0_0008_state::drawgfxzoom_with_pixel_op(gfx_element *gfx, BitmapType &dest, const rectangle &cliprect, uint32_t code, int flipx, int flipy, int32_t destx, int32_t desty, uint32_t scalex, uint32_t scaley, FunctionClass pixel_op)
+{
+	do {
+		assert(dest.valid());
+		assert(dest.cliprect().contains(cliprect));
+
+		// ignore empty/invalid cliprects
+		if (cliprect.empty())
+			break;
+
+		// compute scaled size
+		uint32_t dstwidth = (scalex * gfx->width() + 0x8000) >> 16;
+		uint32_t dstheight = (scaley * gfx->height() + 0x8000) >> 16;
+		if (dstwidth < 1 || dstheight < 1)
+			break;
+
+		// compute 16.16 source steps in dx and dy
+		int32_t dx = (gfx->width() << 16) / dstwidth;
+		int32_t dy = (gfx->height() << 16) / dstheight;
+
+		// compute final pixel in X and exit if we are entirely clipped
+		int32_t destendx = destx + dstwidth - 1;
+		if (destx > cliprect.right() || destendx < cliprect.left())
+			break;
+
+		// apply left clip
+		int32_t srcx = 0;
+		if (destx < cliprect.left())
+		{
+			srcx = (cliprect.left() - destx) * dx;
+			destx = cliprect.left();
+		}
+
+		// apply right clip
+		if (destendx > cliprect.right())
+			destendx = cliprect.right();
+
+		// compute final pixel in Y and exit if we are entirely clipped
+		int32_t destendy = desty + dstheight - 1;
+		if (desty > cliprect.bottom() || destendy < cliprect.top())
+			break;
+
+		// apply top clip
+		int32_t srcy = 0;
+		if (desty < cliprect.top())
+		{
+			srcy = (cliprect.top() - desty) * dy;
+			desty = cliprect.top();
+		}
```

Comment:
This function has a bunch of Allman style.  Please change it to K&R to match the rest of the file.